tests/test_validation.py

```diff
@@ -0,0 +1,237 @@
+    assert model.meta.required_keyword == "foo"
+
+    model = RequiredModel(pass_invalid_values=True)
+
+    with pytest.warns(None) as warnings:
+        model.meta.required_keyword = "foo"
+    assert len(warnings) == 0
+
+    with pytest.warns(ValidationWarning):
+        del model.meta.required_keyword
+    assert model.meta.required_keyword is None
+
+
+def test_pass_invalid_values_attribute_assignment(monkeypatch):
+    def assert_passed(model, passed):
+        with pytest.warns(ValidationWarning):
+            model.meta.string_attribute = 42
+        if passed:
+            assert model.meta.string_attribute == 42
+        else:
+            assert model.meta.string_attribute is None
+
+    assert_passed(ValidationModel(), False)
+    assert_passed(ValidationModel(pass_invalid_values=True), True)
+    assert_passed(ValidationModel(pass_invalid_values=False), False)
+
+    monkeypatch.setenv("PASS_INVALID_VALUES", "true")
+    assert_passed(ValidationModel(), True)
+
+    monkeypatch.setenv("PASS_INVALID_VALUES", "false")
+    assert_passed(ValidationModel(), False)
+
+
+@pytest.mark.xfail(reason="validation on write not yet implemented for ASDF files", strict=True)
+def test_pass_invalid_values_on_write(tmp_path):
+    file_path = tmp_path/"test.asdf"
+    model = ValidationModel(pass_invalid_values=True)
+    with pytest.warns(ValidationWarning):
+        model.meta.string_attribute = 42
+    with pytest.warns(ValidationWarning):
+        model.save(file_path)
+
+    with asdf.open(file_path) as af:
+        assert af["meta"]["string_attribute"] == 42
+
+
+def test_strict_validation_attribute_assignment(monkeypatch):
+    def assert_strict(model, strict):
+        if strict:
+            with pytest.raises(ValidationError):
+                model.meta.string_attribute = 42
+            assert model.meta.string_attribute is None
+        else:
+            with pytest.warns(ValidationWarning):
+                model.meta.string_attribute = 42
+            assert model.meta.string_attribute is None
+
+    assert_strict(ValidationModel(), False)
+    assert_strict(ValidationModel(strict_validation=True), True)
+    assert_strict(ValidationModel(strict_validation=False), False)
+
+    monkeypatch.setenv("STRICT_VALIDATION", "true")
+    assert_strict(ValidationModel(), True)
+
+    monkeypatch.setenv("STRICT_VALIDATION", "false")
+    assert_strict(ValidationModel(), False)
+
+
+def test_validate():
+    model = ValidationModel(pass_invalid_values=True)
+
+    with pytest.warns(None) as warnings:
+        model.meta.string_attribute = "foo"
+        model.validate()
+
+    with pytest.warns(ValidationWarning):
+        model.meta.string_attribute = 42
+    assert model.meta.string_attribute == 42
+
+    with pytest.warns(ValidationWarning):
+        model.validate()
+
+
+def test_validate_required_fields():
+    model = FitsRequiredModel()
+
+    with pytest.warns(ValidationWarning):
+        model.validate_required_fields()
```

Comment:
Is this an existing function in `jwst.datamodels`?  Seems odd to have a separate validator for `fits_required`.

Reply:
Yes, 'tis!  I have no idea why this is necessary.  I'll add an issue to investigate removing it:

https://github.com/spacetelescope/stdatamodels/issues/12